File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_MESSAGEBOX.cs
﻿using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_MESSAGEBOX : Message
    {
        public SID_MESSAGEBOX()
        {
            Id = (byte)MessageIds.SID_MESSAGEBOX;
            Buffer = new byte[6];
        }

        public SID_MESSAGEBOX(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_MESSAGEBOX;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

            if (context.Direction != MessageDirection.ServerToClient)
                throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be sent from server to client");

            if (Buffer.Length < 6)
                throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 6 bytes");

            var style = (UInt32)context.Arguments["style"];
            var text = (string)context.Arguments["text"];
            var caption = (string)context.Arguments["caption"];

            Buffer = new byte[6 + Encoding.UTF8.GetByteCount(text) + Encoding.UTF8.GetByteCount(caption)];

            using var m = new MemoryStream(Buffer);
            using var w = new BinaryWriter(m);

            w.Write((UInt32)style);
            w.Write((string)text);
            w.Write((string)caption);

            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Sent MessageBox parameters to client");
            context.Client.Send(ToByteArray(context.Client.ProtocolType));
            return true;
        }

        public new byte[] ToByteArray(ProtocolType protocolType)
        {
            if (protocolType.IsChat())
            {
                using var _m = new MemoryStream(Buffer);
                using var r = new BinaryReader(_m);
                var style = r.ReadUInt32();
                var text = r.ReadByteString();
                var caption = r.ReadByteString();

                using var m = new MemoryStream();
                using var w = new System.IO.BinaryWriter(m);
                w.Write(Encoding.UTF8.GetBytes($"{2000 + Id} MESSAGEBOX \""));
                w.Write(text);
                w.Write((byte)'"');
                w.Write(Encoding.UTF8.GetBytes(Battlenet.Common.NewLine));
                return m.GetBuffer()[0..(int)w.BaseStream.Length];
            }
            else
            {
                return base.ToByteArray(protocolType);
            }
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_MESSAGEBOX.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_MESSAGEBOX
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_MESSAGEBOX)
            Buffer = New Byte(5) {}
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_MESSAGEBOX)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
            If context.Direction <> MessageDirection.ServerToClient Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be sent from server to client")
            If Buffer.Length < 6 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 6 bytes")
            Dim style = CUInt(context.Arguments("style"))
            Dim text = CStr(context.Arguments("text"))
            Dim caption = CStr(context.Arguments("caption"))
            Buffer = New Byte(6 + Encoding.UTF8.GetByteCount(text) + Encoding.UTF8.GetByteCount(caption) - 1) {}

            Using m = New MemoryStream(Buffer)
                Using w = New BinaryWriter(m)
                    w.Write(CUInt(style))
                    w.Write(CStr(text))
                    w.Write(CStr(caption))
                End Using
            End Using

            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
            context.Client.Send(ToByteArray(context.Client.ProtocolType))
            Return True
        End Function
    End Class
End Namespace
